tools/topology/sof/pipe-codec-adapter-capture.m4

```diff
@@ -0,0 +1,139 @@
+# Low Latency Passthrough Pipeline and PCM
+#
+# Pipeline Endpoints for connection are :-
+#
+#  host PCM_C <-- B0 <-- CODEC_ADAPTER <-- B1 <-- sink DAI0
+
+# Include topology builder
+include(`utils.m4')
+include(`buffer.m4')
+include(`pcm.m4')
+include(`dai.m4')
+include(`pipeline.m4')
+include(`codec_adapter.m4')
+include(`bytecontrol.m4')
+
+#
+# Controls
+#
+
+# For codec developers, please define the following control bytes of setup config and runtime params
+# and the corresponding max bytenum on your own.
+
+# Codec Adapter setup config control bytes (little endian)
+#  : bytes "abi_header, ca_config, [codec_param0, codec_param1...]"
+#  - 32 bytes abi_header: you could get by command "sof-ctl -t 0 -g <payload_size> -b"
+#    - [0:3]: magic number 0x00464f53
+#    - [4:7]: type 0
+#    - [8:11]: payload size in bytes (not including abi header bytes)
+#    - [12:15]: abi 3.1.0
+#    - [16:31]: reserved 0s
+#  - 20 bytes ca_config: codec adapter setup config parameters, for more details please refer
+#                        struct ca_config under audio/codec_adapter/codec/generic.h
+#    - [0]: API ID, e.g. 0x01
+#    - [1:3]: codec ID, e.g. 0xd03311
+#    - [4:7]: reserved 0s
+#    - [8:11]: sample rate, e.g. 48000
+#    - [12:15]: sample width in bits, e.g. 32
+#    - [16:19]: channels, e.g. 2
+# - (optional) 12+ bytes codec_param: codec TLV parameters container, for more details please refer
+#                                     struct codec_param under audio/codec_adapter/codec/generic.h
+#    - [0:3]: param ID
+#    - [4:7]: size in bytes (ID + size + data)
+#    - [8:n-1]: data[], the param data
+ifdef(`CA_SETUP_CONTROLBYTES',`', `define(`CA_SETUP_CONTROLBYTES',
```

Comment:
random question, how does this work if two adapter pipelines are included? shouldnt we just implicitly unset this and reset it?

Reply:
good question. I think we should undefine them in the end of pipe-codec-adapter-[capture|playback].m4 to void this issue.

Reply:
given you have now unset, isn't ifdef unneeded?

Reply:
These define variables are used for a codec pipeline to overwrite. For example, pipe-processing-playback.m4 defines its own CA_SETUP_CONTROLBYES, and CA_RUNTIME_CONTROLBYTES, and others.
"ifdef" here is to set a default value.
And in order not to re-define variables in the cases of multiple codecs, I chose to "undefine" in the end of pipe-codec-adapter-playback.m4

Reply:
Yes, im just suggesting we can drop the `if` since we are unsetting above and just put a `define` here since it will always be undefined.

Reply:
Sorry could you deliberate more on "since we are unsetting above"? I still didn't get what is unsetting?

Reply:
ah i misread the code on my last pass, I thought you were doing this
```
unset
ifdef
```
which means that our namespace should always be cleared. I think it might be better assuming unsetting an undefined does not result in an error